### ISSUE TITLE
`code new_file.ipynb` and then saving does not always honour the path of the new file

### ISSUE THREAD

**Issue by starball5** [microsoft/vscode#190703]
I tried searching for duplicates by googling `github vscode jupyter issues save new file created from terminal wrong path` and looking through the first few pages of search results.

Does this issue occur when all extensions are disabled?: Yes (I tried with `code --disable-extensions new_file.ipynb`)

- VS Code Version: 1.81.1
- OS Version: Ubuntu 22.04.2

Steps to Reproduce:

1. navigate a shell to some directory A
2. run `code --disable-extensions new_file.ipynb`
3. quit VS Code without saving
4. navigate the same shell to some directory B
5. run `code --disable-extensions new_file.ipynb`
6. check that the path of the file is anchored to the cwd of the shell that ran the previous command when the command was run by running `File: Copy Path of Active File` in the command palette
7. save the file
8. observe in the breadcrumbs that the file is _no longer_ anchored to that directory, but was saved to some other location.

For me, that location is my user home directory.

This behaviour was brought to my attention by [this Stack Overflow question](https://stackoverflow.com/q/76917125/11107541), which is technically a duplicate of [this older Stack Overflow question](https://stackoverflow.com/q/71924488/11107541).

**Comment by starball5** [comment]
@bpasero nope. I get the behaviour that I've already described and not the one that you get. And it's not only reflected in the breadcrumbs after saving, but also by `File: Copy Path of Active File`. So the result of `File: Copy Path of Active File` differs before and after saving. But isn't the behaviour you described a bug as well? Just a different one? What's your version info? Mine is this:

```none
Version: 1.81.1
Commit: 6c3e3dba23e8fadc360aed75ce363ba185c49794
Date: 2023-08-09T22:18:39.991Z
Electron: 22.3.18
ElectronBuildId: 22689846
Chromium: 108.0.5359.215
Node.js: 16.17.1
V8: 10.8.168.25-electron.0
OS: Linux x64 6.2.0-26-generic snap
```

**Comment by bpasero** [comment]
Can you try to reproduce with our nightly insider builds? You can give our preview releases a try from: https://code.visualstudio.com/insiders/

**Comment by starball5** [comment]
@bpasero I just tried on Insiders and got the same behaviour that I described in my comments above.

```none
Version: 1.82.0-insider
Commit: 501baeb5e07b5a404f862d922bacb93771d91331
Date: 2023-08-18T05:32:21.521Z
Electron: 25.5.0
ElectronBuildId: 23084831
Chromium: 114.0.5735.289
Node.js: 18.15.0
V8: 11.4.183.29-electron.0
OS: Linux x64 6.2.0-26-generic snap
```

**Comment by bpasero** [comment]
I can also not reproduce on my Ubuntu 22, are you doing anything different?

![Recording 2023-08-22 at 14 38 55](https://github.com/microsoft/vscode/assets/900690/825d4b86-b86c-4184-aae2-a32978e59384)

**Comment by starball5** [comment]
@bpasero I was actually closing the first one before exiting VS Code (which I forgot to say in my repro steps (sorry)), but that doesn't seem to make a difference. Even if I don't close the first one and allow it to be recalled when opening the second one, saving the second one still saves it to a wrong location. Interesting thing: if I keep the first one around, invoking save once the second one is opened opens the system dialog to select a save location.

**Comment by bpasero** [comment]
Oh nice, I can reproduce, thanks!

@rebornix @amunger there are actually 2 bugs here. 

**Wrong Path**
* close all instances of code
* `cd` into a meaningful folder such as `~/foo`
* from the command line `code new.ipynb`
* save the notebook

=>  🐛 you are not being asked for a location but the location ends up being `~/new.ipynb` and not `~/foo/new.ipynb`

The issue seems to be here where you take the `pathCandidate` even though you have an associated path:

https://github.com/microsoft/vscode/blob/ad94cf18189746d2e942fda257950901049b3af3/src/vs/workbench/contrib/notebook/common/notebookEditorInput.ts#L197-L199

Here is a code snippet for how an untitled text model converts an associated path to a good local resource based on the environment:

https://github.com/microsoft/vscode/blob/ad94cf18189746d2e942fda257950901049b3af3/src/vs/workbench/services/textfile/browser/textFileService.ts#L584-L587

**Lost Associated Resource**
* close all instances of code
* `cd` into a meaningful folder such as `~/foo`
* from the command line `code new.ipynb`
* reload the window
* save the notebook

=> 🐛 you are being asked for a location via a file dialog

This is because when resolving the notebook after the reload, the working copy ends up to loose its associated file path. I think the issue is somewhere here:

https://github.com/microsoft/vscode/blob/ad94cf18189746d2e942fda257950901049b3af3/src/vs/workbench/contrib/notebook/common/notebookEditorModelResolverServiceImpl.ts#L179

Here is how the untitled editor service tries to figure out if the untitled editor has an associated path or not. It does so by purely checking for the resource path being `Untitled-<number>` or something else via `isUntitledWithAssociatedResource`:

https://github.com/microsoft/vscode/blob/ad94cf18189746d2e942fda257950901049b3af3/src/vs/workbench/services/untitled/common/untitledTextEditorHandler.ts#L116-L120

Its not a 100% perfect solution but avoids having to store another property.
